### PR TITLE
[Renaming] Handle crash on FuncCall with Arg class name conflict with use statement or namespace on RenameClassRector with auto import enabled

### LINKED ISSUE #7206: RenameClassRector crashes because multiple import of same class(?)
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.13.4             |

While reporting https://github.com/rectorphp/rector/issues/7205 I noticed a crash.

I happens because of this line:
```
use CompanyB\Request;
```

RenameClassRector tries to add a new import with the same name. 

## Minimal PHP Code Causing Issue

https://getrector.org/demo/4d566330-9ae1-49fd-b1bb-8cbae0e9c25e

## Expected Behaviour

Not crash. I guess fixing #7205 would solve it. But just wanted to report it in case it can be handled in a different way.

### PR DESCRIPTION
**Pull request by samsonasik**
Fixes https://github.com/rectorphp/rector/issues/7206

It currently crash:

```bash
There were 2 failures:

1) Rector\Tests\Renaming\Rector\Name\RenameClassRector\AutoImportNamesTest::test with data set #10 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
assert($functionInfo instanceof \PHPStan\BetterReflection\Reflection\ReflectionFunction || $functionInfo === null) in phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflector/DefaultReflector.php:59

2) Rector\Tests\Renaming\Rector\Name\RenameClassRector\AutoImportNamesTest::test with data set #2 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
assert($functionInfo instanceof \PHPStan\BetterReflection\Reflection\ReflectionFunction || $functionInfo === null) in phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflector/DefaultReflector.php:59
```

### CODE REVIEW

**Comment by samsonasik** [comment]
Fixed 🎉 /cc @eigan

**Comment by samsonasik** [comment]
The phpstan notice is unrelated:

```bash
  src/NodeManipulator/PropertyManipulator.php:261
 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  - '#On passing a constant, the method should have an enum type\. See https\://phpstan\.org/writing\-php\-code/phpdoc\-types\#literals\-and\-constants#'
 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  src/Reflection/ReflectionResolver.php:207
 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  - '#On passing a constant, the method should have an enum type\. See https\://phpstan\.org/writing\-php\-code/phpdoc\-types\#literals\-and\-constants#'
 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


                                                                                                                        
 [ERROR] Found 55 errors        
```

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php [hunk 1]
```diff
@@ -27,6 +27,12 @@ public function __construct(
 
     public function locateIdentifier(Reflector $reflector, Identifier $identifier): ?Reflection
     {
+        $identifierType = $identifier->getType();
+        $typeName = $identifierType->getName();
+        if ($typeName !== 'PHPStan\BetterReflection\Reflection\ReflectionClass') {
+            return null;
+        }
+
```
Comment (TomasVotruba) [inline src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: It looks very weird. What is this related to?

Comment (samsonasik) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: it is happen when renamed class is inside func call, eg:

```php
app(Queue::class);
```

without this, it will crash: 


```bash
There were 2 failures:

1) Rector\Tests\Renaming\Rector\Name\RenameClassRector\AutoImportNamesTest::test with data set #10 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
assert($functionInfo instanceof \PHPStan\BetterReflection\Reflection\ReflectionFunction || $functionInfo === null) in phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflector/DefaultReflector.php:59

2) Rector\Tests\Renaming\Rector\Name\RenameClassRector\AutoImportNamesTest::test with data set #2 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
assert($functionInfo instanceof \PHPStan\BetterReflection\Reflection\ReflectionFunction || $functionInfo === null) in phar:///Users/samsonasik/www/rector-src/vendor/phpstan/phpstan/phpstan.phar/vendor/ondrejmirtes/better-reflection/src/Reflector/DefaultReflector.php:59

Comment (TomasVotruba) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: We should figure out why this happens. This ignore seem to hard coded solution.

What about fixture with method call?

Comment (samsonasik) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: in method call arg seems working ok https://github.com/rectorphp/rector-src/pull/2431/commits/525c3479acb3dadbf4d60a3317227d5eb22b2095 .

Comment (samsonasik) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: The identifier type is `PHPStan\BetterReflection\Reflection\ReflectionFunction` if not class:

```bash
.PHPStan\BetterReflection\Identifier\IdentifierType #17599
   name: 'PHPStan\BetterReflection\Reflection\ReflectionFunction'
```

we can exclude it for `PHPStan\BetterReflection\Reflection\ReflectionFunction` only if we want to avoid the negation check:

```diff
-        if ($typeName !== 'PHPStan\BetterReflection\Reflection\ReflectionClass') {
+        if ($typeName === 'PHPStan\BetterReflection\Reflection\ReflectionFunction') {
```

so we can add other type name if there is other use case like this again. what do you think?

Comment (samsonasik) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: @TomasVotruba I found it 🎉 , it seems resolved by verify start line < 0: 

```diff
-        return ReflectionClass::createFromNode($classReflector, $class, $fakeLocatedSource);
+        $reflectionClass = ReflectionClass::createFromNode($classReflector, $class, $fakeLocatedSource);

+        if ($reflectionClass->getStartLine() < 0) {
+            return null;
+        }

+        return $reflectionClass;
```

so no need the hard code check https://github.com/rectorphp/rector-src/pull/2431/commits/b126df4cb80db4ee8dbf4e695b1eab7d535377c3

Comment (samsonasik) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: It seems there is error that only in CI:

```diff
There was 1 failure:
1) Rector\Tests\Renaming\Rector\Name\RenameClassRector\RenameClassRectorTest::test with data set #13 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
rules-tests/Renaming/Rector/Name/RenameClassRector/Fixture/rename_class.php.inc
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
 class MyNewClass
 {
     /**
-     * @return \MyNewNamespace\MyNewClass
+     * @return MyClass
```

Comment (samsonasik) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: I think the way to go is compare equal to `PHPStan\BetterReflection\Reflection\ReflectionFunction` so it should be safe as this class is checking renamed class, not renamed function https://github.com/rectorphp/rector-src/pull/2431/commits/4fe4bc79c1f51d8739f18728de31d7cccc7fb7b9

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

File: src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php (hunk 1 above)
Comment (samsonasik) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: @TomasVotruba it seems `RenamedClassesSourceLocator` can be completely removed since PHPStan 1.7.x is fully static reflection, let's remove it https://github.com/rectorphp/rector-src/pull/2431/commits/097753b1e742345bb3d3f460aa5719b2f4aa0a19

Comment (samsonasik) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: Back to get error in CI : 

```diff
There was 1 failure:
1) Rector\Tests\Renaming\Rector\Name\RenameClassRector\RenameClassRectorTest::test with data set #13 (Symplify\SmartFileSystem\SmartFileInfo Object (...))
rules-tests/Renaming/Rector/Name/RenameClassRector/Fixture/rename_class.php.inc
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
 class MyNewClass
 {
     /**
-     * @return \MyNewNamespace\MyNewClass
+     * @return MyClass
```

Comment (samsonasik) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: I reverted the removal of `RenamedClassesSourceLocator`, back to use compare type name is `PHPStan\BetterReflection\Reflection\ReflectionFunction` https://github.com/rectorphp/rector-src/pull/2431/commits/4b1cbd4b2dd50dc1a215a107a5743f4f1ad70bb7

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

File: src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php (hunk 1 above)
Comment (samsonasik) [reply src/StaticReflection/SourceLocator/RenamedClassesSourceLocator.php]: @TomasVotruba ok, I found it 🎉 , it seems resolved by check : 

```php
        if (! $identifier->isClass()) {
            return null;
        }
```

the check `isClass()` is check against `\PHPStan\BetterReflection\Reflection\ReflectionClass` type which check even on `interface`.

https://github.com/rectorphp/rector-src/blob/b1583c7d8636aa32e6d5f7bdf37162e3e38a3907/rules-tests/Renaming/Rector/Name/RenameClassRector/config/configured_rule.php#L42-L43

**Comment by samsonasik** [comment]
Finally, no hard code check 🎉🎉🎉 https://github.com/rectorphp/rector-src/pull/2431/commits/591ce37f92d9f18d0a4fb5a48a38119e781b3c8f  , All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by samsonasik** [comment]
rebased.

**Comment by TomasVotruba** [comment]
Thank you :+1:
